p2p/tpu.md

```diff
@@ -0,0 +1,234 @@
+# TPU protocol
+
+TPU is a publicly available peer-to-peer service to queue transactions for inclusion on the Solana blockchain.
+
+It is backed by a streaming protocol (UDP datagrams or QUIC) which routes transactions to network leaders.
+
+## Topology
+
+- **Clients** produce new transactions
+- **Relayers** forward transactions to leaders or other relayers
+- **Leaders** pack transactions into blocks
+
+The lifecycle of a transaction starts with the signer.
+Signers can either submit transactions directly to TPU nodes (requires UDP connectivity) or submit them through RPC servers.
+
+```mermaid
+flowchart LR
+
+subgraph Clients
+  C1[Client]
+  C2[Client]
+  C3[Client]
+  C4[Client]
+end
+
+subgraph Validators
+  L[Leader]
+  V1[Validator]
+  V2[Validator]
+end
+
+R1[Node]
+
+C1 -- RPC --> R1
+C2 -- TPU --> R1
+R1 -- TPU --> L
+C3 -- TPU --> L
+C4 -- TPU ---> V1
+V1 -- TPU\nTPUvote --> L
+V2 -- TPUvote --> L
+```
+
+In a Solana cluster, full nodes usually participate as TPU relayers, although not required.
+All validators included in the leader schedule should operate an open TPU service.
+
+Nodes attempt to forward transactions to the current leader's TPU endpoints.
+The node identities of the leader schedule are well known (runtime intrinsic),
+and their TPU endpoints are part of CRDS ContactInfo.
+
+## Classes
+
+The network uses separate classes of TPU traffic for QoS.
+
+- The **TPUvote** class is Tower BFT consensus messages (highest priority).
+- The default **TPU** class is regular user transactions, generated by validators.
+- The **TPUfwd** class is user transactions that the previous leader did not fully process.
+
+## TPU/QUIC protocol v1
+	
+TPU/QUIC relays transactions over the QUIC transport protocol.
+
+QUIC is defined by [RFC 9000] and runs on UDP networks.
+
+Data is encrypted using TLS 1.3, defined by [RFC 8446].
+
+  [RFC 8446]: https://www.rfc-editor.org/rfc/rfc8446.html
+  [RFC 9000]: https://www.rfc-editor.org/rfc/rfc9000.html
+
```

Comment:
idk if worth mentioning here, but quic port is current tpu_port + 6

Reply:
The destination port numbers are irrelevant in this context since nodes always discover them from gossip.

But good point, we should probably add:
- TPU/UDP and TPU/QUIC cannot be multiplexed on the same port
- The source port number is arbitrary but needs to stay stable for TPU/QUIC

Reply:
true, the point was more about the tpu currently advertised in gossip is udp and quic is +6 from that.

Reply:
Ah, so the TPU/QUIC port is implied not explicit? Let's document this in the gossip protocol pages

Reply:
Fixed, adding the +6 quirk into gossip pages.